### PR TITLE
CDTT-471 - Integrating the Google Cloud Storage functionality; To run…

### PR DESCRIPTION
**Pull request by dragotaa**
This Jira item is related to integration of the Google Cloud Storage into Census RH Service.
This integration allows RH Service to write in a predefined bucket objects containing UAC and CASE details.

### CODE REVIEW

**Comment by philwhiles** [comment]
- CloudStorage.create() should be .store() as we will call it to either create or update an object.
- caseId not caseID (the id part is an abbreviation not an acronym, and Id would be consistent with RM code convention)
- does the exception caught by the create() only occur when the bucket already exists? what other errors might result in an exception? if other reasons for exception, we need to examine cause and possibly rethrow.
- we need to create two model types, even if flat for time being to represent the Case and the UAC details, see JIRA for 
initial stab at contents. These should contain typed fields, and the Cloud interface should accept and return these, rather than any old String. 
- The cloud implies should take the typed objects and serialise them as JSON to be stored.
- We do not have a delete use case at this time - we don't want to have unused code. Every line of code we write is code to maintain and a source of bugs!
- We do however have a read use case
- Re tests - these are integration and not unit tests if they have to go out to the cloud. They should be suffixed with 'IT' and we will need to make sure the maven test plugin knows to exclude them from the unit test phase.
- We will need to ensure we have a test account in google by the sounds of it

File: src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/DataInCloudServiceIT_Test.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        caseContext.setAddress(address);
+        caseContext.setTimestamp(timestamp);
+        caseContext.setCollectionExercise("collection exercise");
+        caseContext.setPostcode("PO155XX");
+        caseContext.setUprn("123456789012");
+        caseContext.setRegion(region);
+    }
+
+    @Test
+    public void writeAndReadUAC() {
```
Comment (philwhiles) [inline src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/DataInCloudServiceIT_Test.java]: You see - you had your java developer hat on when you wrote this one!

File: src/test/java/uk/gov/ons/ctp/integration/rhsvc/cloud/GCSServiceIT_Test.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
+package uk.gov.ons.ctp.integration.rhsvc.cloud;
+
+import org.junit.Assert;
+import org.junit.FixMethodOrder;
+import org.junit.Test;
+import org.junit.runners.MethodSorters;
+
+@FixMethodOrder(MethodSorters.NAME_ASCENDING)
+public class GCSServiceIT_Test {
```
Comment (philwhiles) [inline src/test/java/uk/gov/ons/ctp/integration/rhsvc/cloud/GCSServiceIT_Test.java]: **This currently prevents a mvn clean install due to the need to external service.**
Please merge in master - I have added skip integration tests to pom.
Then rename this class to end just '_IT' and it will be skipped unless we provide the right arg to mvn

File: src/test/java/uk/gov/ons/ctp/integration/rhsvc/cloud/GCSServiceIT_Test.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
+package uk.gov.ons.ctp.integration.rhsvc.cloud;
```
Comment (philwhiles) [inline src/test/java/uk/gov/ons/ctp/integration/rhsvc/cloud/GCSServiceIT_Test.java]: We need to maintain code coverage, and also test all routes through code, so please add some tests that assert null handling as mentioned also

File: src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/DataInCloudServiceIT_Test.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    CaseContext caseContext;
+
+    @Before
+    public void setup(){
+        storage = new GCSService();
+        cloud = new DataInCloudService(storage);
+        caseId = "123456789";
+        uacCode = "abcd-defg-1234";
+
+        String case_type = "(H|HI|C|CI)";
```
Comment (philwhiles) [inline src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/DataInCloudServiceIT_Test.java]: caseType will be only one of those indicated by the regex ie "H"

File: src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/DataInCloudServiceIT_Test.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        try {
+            cloud.writeObject(caseContext);
+        } catch (JsonProcessingException e) {
+            e.printStackTrace();
+            fail("Failed to write upload case object to cloud storage");
+        }
+        CaseContext caseContext2 = null;
+        try {
+            caseContext2 = cloud.readCase(caseId);
+        } catch (IOException e) {
```
Comment (philwhiles) [inline src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/DataInCloudServiceIT_Test.java]: As a pattern it is OK to allow the underlying readCase to throw and not catch - change the method sig to throw exception, and test will still fail - readCase should log why it failed

less code in the test then

File: src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/DataInCloudServiceIT_Test.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+            e.printStackTrace();
+            fail("Failed to read case object from cloud storage");
+        }
+        assertEquals(caseContext.getCaseType(), caseContext2.getCaseType());
```
Comment (philwhiles) [inline src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/DataInCloudServiceIT_Test.java]: Try using equals() as provided by Lombok instead of all these asserts

File: src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/DataInCloudServiceIT_Test.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        caseContext.setPostcode("PO155XX");
+        caseContext.setUprn("123456789012");
+        caseContext.setRegion(region);
+    }
+
+    @Test
+    public void writeAndReadUAC() {
+        try {
+            cloud.writeObject(uac);
+        } catch (JsonProcessingException e) {
```
Comment (philwhiles) [inline src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/DataInCloudServiceIT_Test.java]: ditto

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/DataInCloud.java [hunk 8]
```diff
@@ -0,0 +1,14 @@
+package uk.gov.ons.ctp.integration.rhsvc.service;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import uk.gov.ons.ctp.integration.rhsvc.domain.model.CaseContext;
+import uk.gov.ons.ctp.integration.rhsvc.domain.model.UACContext;
+
+import java.io.IOException;
+
+public interface DataInCloud {
```
Comment (philwhiles) [inline src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/DataInCloud.java]: Ella had already created a RespondentDataService for this purpose - lets move this functionality into that

File: pom.xml [hunk 9]
```diff
@@ -80,7 +80,7 @@
             <artifactId>spring-boot-starter-actuator</artifactId>
         </dependency>
 
-	<dependency>
+        <dependency>
             <groupId>org.springframework</groupId>
             <artifactId>spring-aspects</artifactId>
             <version>4.3.8.RELEASE</version>
```
[the remark this answers is not in the document]
Comment (dragotaa) [reply pom.xml]: I did not change anything in there. I will check and if there is any change required in there I will set it up in the BOM

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/cloud/CloudService.java [hunk 10]
```diff
@@ -0,0 +1,8 @@
+package uk.gov.ons.ctp.integration.rhsvc.cloud;
+
+public interface CloudService {
```
[the remark this answers is not in the document]
Comment (dragotaa) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/cloud/CloudService.java]: Will change it

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/cloud/GCSService.java [hunk 11]
```diff
@@ -0,0 +1,68 @@
+package uk.gov.ons.ctp.integration.rhsvc.cloud;
+
+import com.godaddy.logging.Logger;
+import com.godaddy.logging.LoggerFactory;
+import com.google.cloud.storage.*;
+import org.springframework.stereotype.Service;
+
+@Service
+public class GCSService implements CloudService {
```
[the remark this answers is not in the document]
Comment (dragotaa) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/cloud/GCSService.java]: Done

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/DataInCloud.java [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+    void writeObject(UACContext uac) throws JsonProcessingException;
```
[the remark this answers is not in the document]
Comment (dragotaa) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/DataInCloud.java]: Done

File: src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/DataInCloudServiceIT_Test.java (hunk 4 above)
Comment (dragotaa) [reply src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/DataInCloudServiceIT_Test.java]: Done

File: src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/DataInCloudServiceIT_Test.java (hunk 5 above)
Comment (dragotaa) [reply src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/DataInCloudServiceIT_Test.java]: Done

File: src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/DataInCloudServiceIT_Test.java (hunk 7 above)
Comment (dragotaa) [reply src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/DataInCloudServiceIT_Test.java]: done

File: src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/DataInCloudServiceIT_Test.java (hunk 6 above)
Comment (dragotaa) [reply src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/DataInCloudServiceIT_Test.java]: We can talk about this on Monday.
